specs/interop.md

```diff
@@ -0,0 +1,324 @@
+where the user can send a relaying message before the initiating message and trick the derivation pipeline
+into reorganizing the sequencer.
+
+### Safety
+
+The initiating messages for all relaying messages MUST be resolved as safe before an L2 block can transition from being unsafe to safe.
+Users MAY optimistically accept unsafe blocks without any verification of the relaying messages. They SHOULD optimistically verify
+the initiating messages exist in remote unsafe blocks to more quickly reorganize out invalid blocks.
+
+## State Transition Function
+
+After the full execution of a block, the set of `MessagePassed` events emitted from the `CrossL2Outbox` are collected and merklized.
+This commitment MUST be included as the block header's [extra data field](https://github.com/ethereum/execution-specs/blob/1fed0c0074f9d6aab3861057e1924411948dc50b/src/ethereum/frontier/fork_types.py#L115).
+The events are serialized with using [Simple Serialize](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md) aka SSZ.
+
+| Name          | Value |
+|---------------|-------------------------------------|
+| `MAX_LOG_DATA_SIZE` | `uint64(2**24)` or 16,777,216 |
+| `MAX_MESSAGES_PER_BLOCK` | `uint64(2**20)` or 1,048,576 |
+
+```python
+class CrossChainMessage(Container):
+    nonce: uint256
+    sender: ExecutionAddress
+    target: ExecutionAddress
+    value: uint256
+    timestamp: uint256
+    chainid: uint256
+    data: ByteList[MAX_LOG_DATA_SIZE]
+
+messages = List[CrossChainMessage, MAX_MESSAGES_PER_BLOCK]
+
+block.extra_data = messages.hash_tree_root()
+```
+
+## Predeploys
+
+Two new system level predeploys are introduced for managing cross chain messaging.
+
+### CrossL2Outbox
+
+Address: `0x4200000000000000000000000000000000000023`
+
+The `CrossL2Outbox` is responsible for initiating a cross chain message on the local chain.
+For every cross chain message that is sent, it MUST emit an event that indicates the existence
+of the cross chain message.
+
+The event SHOULD make it as cheap as possible for the remote chain to look up the existence
+of the initiating transaction when verifying the relaying transaction.
+
+#### Initiating Messages
+
+An initiating message emits the following event:
+
+```solidity
+event MessagePassed(
+    bytes32 indexed messageHash,
+    address indexed sender,
+    address indexed target,
+    uint256 nonce,
+    uint256 value,
+    uint256 timestamp,
+    uint256 chainid,
+    bytes data
+);
+```
+
+The `messageHash` and chain id uniquely identify the existence of an initiating message and SHOULD
+be used by a block builder to ensure the existence of the initiating message when verifying the relaying
+transaction. `messageHash` SHOULD be `indexed` to allow for easy filtering of logs.
+
+A cross chain message is defined in the following format:
+
+```solidity
+struct CrossChainMessage {
+    uint256 nonce;
```

Comment:
There are 2 ways to do upgradability with this. One is to introduce a `CrossChainMessageV2` and another is to put the version in the nonce. Another possibility is to do something like:

```solidity
struct CrossChainmessage {
  uint256 version;
  bytes data;
}
```

Where the version defines the encoding of the data